File: main/src/main/java/com/sedmelluq/discord/lavaplayer/tools/io/AbstractHttpInterfaceManager.java
package com.sedmelluq.discord.lavaplayer.tools.io;

import org.apache.http.client.config.RequestConfig;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClientBuilder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.function.Consumer;
import java.util.function.Function;

/**
 * Base class for an HTTP interface manager with lazily initialized http client instance.
 */
public abstract class AbstractHttpInterfaceManager implements HttpInterfaceManager {
  private static final Logger log = LoggerFactory.getLogger(AbstractHttpInterfaceManager.class);

  private final HttpClientBuilder clientBuilder;
  private final Object lock;
  private boolean closed;
  private CloseableHttpClient sharedClient;
  private RequestConfig requestConfig;

  /**
   * @param clientBuilder HTTP client builder to use for creating the client instance.
   * @param requestConfig Request config used by the client builder
   */
  public AbstractHttpInterfaceManager(HttpClientBuilder clientBuilder, RequestConfig requestConfig) {
    this.clientBuilder = clientBuilder;
    this.requestConfig = requestConfig;
    this.lock = new Object();
  }

  @Override
  public void close() throws IOException {
    synchronized (lock) {
      closed = true;

      if (sharedClient != null) {
        CloseableHttpClient client = sharedClient;
        sharedClient = null;
        client.close();
      }
    }
  }

  @Override
  public void configureRequests(Function<RequestConfig, RequestConfig> configurator) {
    synchronized (lock) {
      try {
        close();
      } catch (Exception e) {
        log.warn("Failed to close HTTP client.", e);
      }

      closed = false;
      requestConfig = configurator.apply(requestConfig);
      clientBuilder.setDefaultRequestConfig(requestConfig);
    }
  }

  @Override
  public void configureBuilder(Consumer<HttpClientBuilder> configurator) {
    synchronized (lock) {
      try {
        close();
      } catch (Exception e) {
        log.warn("Failed to close HTTP client.", e);
      }

      closed = false;
      configurator.accept(clientBuilder);
    }
  }

  protected CloseableHttpClient getSharedClient() {
    synchronized (lock) {
      if (closed) {
        throw new IllegalStateException("Cannot get http client for a closed manager.");
      }

      if (sharedClient == null) {
        sharedClient = clientBuilder.build();
      }

      return sharedClient;
    }
  }
}

File: main/src/main/java/lavaplayer/tools/io/AbstractHttpInterfaceManager.kt
package lavaplayer.tools.io

import mu.KotlinLogging
import org.apache.http.client.config.RequestConfig
import org.apache.http.impl.client.CloseableHttpClient
import org.apache.http.impl.client.HttpClientBuilder
import java.io.IOException

/**
 * Base class for an HTTP interface manager with lazily initialized http client instance.
 * @param clientBuilder HTTP client builder to use for creating the client instance.
 * @param requestConfig Request config used by the client builder
 */
abstract class AbstractHttpInterfaceManager(
    private val clientBuilder: HttpClientBuilder,
    private var requestConfig: RequestConfig
) : HttpInterfaceManager {
    companion object {
        private val logger = KotlinLogging.logger { }
    }

    private var closed: Boolean = false
    private var sharedClient: CloseableHttpClient? = null

    @Throws(IOException::class)
    override fun close() = synchronized(this) {
        closed = true
        sharedClient?.use { sharedClient = null } ?: Unit
    }

    override fun configureRequests(configurator: RequestConfigurator) {
        synchronized(this) {
            try {
                close()
            } catch (e: Exception) {
                logger.warn(e) { "Failed to close HTTP client." }
            }

            closed = false
            requestConfig = configurator(requestConfig)
            clientBuilder.setDefaultRequestConfig(requestConfig)
        }
    }

    override fun configureBuilder(configurator: BuilderConfigurator) {
        synchronized(this) {
            try {
                close()
            } catch (e: Exception) {
                logger.warn(e) { "Failed to close HTTP client." }
            }

            closed = false
            clientBuilder.apply(configurator)
        }
    }

    protected fun getSharedClient() = synchronized(this) {
        if (closed) {
            throw IllegalStateException("Cannot get http client for a closed manager.")
        }

        if (sharedClient == null) {
            sharedClient = clientBuilder.build()
        }

        sharedClient!!
    }
}
